waku/v2/protocol/waku_store/protocol.nim

```diff
@@ -78,25 +74,68 @@ proc reportStoredMessagesMetric*(w: WakuStore) =
 
   waku_store_messages.set(resCount.value, labelValues = ["stored"])
 
+# TODO: Move to a message store wrapper
+proc isValidMessage(msg: WakuMessage): bool =
+  if msg.timestamp == 0:
+    return true
+
+  let 
+    now = getNanosecondTime(getTime().toUnixFloat())
+    lowerBound = now - MaxMessageTimestampVariance
+    upperBound = now + MaxMessageTimestampVariance
+
+  return lowerBound <= msg.timestamp and msg.timestamp <= upperBound
+
+# TODO: Move to a message store wrapper
+proc handleMessage*(w: WakuStore, pubsubTopic: string, msg: WakuMessage) =
+  if w.store.isNil():
+    # Messages should not be stored
+    return
+
+  if msg.ephemeral:
+    # The message is ephemeral, should not be stored
+    return
+  
+  if not isValidMessage(msg):
+    waku_store_errors.inc(labelValues = [invalidMessage])
+    return
+
 
-proc findMessages(w: WakuStore, query: HistoryQuery): HistoryResponse {.gcsafe.} =
+  let insertStartTime = getTime().toUnixFloat()
```

Comment:
https://github.com/status-im/nwaku/blob/d40636d03be11dcac22d36c3d9aa0dab0fcf97db/waku/v2/utils/time.nim#L28-L34
you can use this for Histograms now

Reply:
Can be addressed in a follow up PR :)

Reply:
All the "message handling and storage" logic, in other words, the logic that is not related to historical message retrieval, will be extracted into a separate entity together with the changes necessary to support the "custom" message validation.

-----

Yes, I agree. Thanks for the heads up. We'll take it when doing the message store split.

JFYI relate to the `nowInUnixFloat` and all the `times` related primitives: There is a piece of work pending to investigate and implement to use `nim-chronos` timers' time primitives to define the `Timestamp` type and all the operations.

Reply:
Okay